Azure observability install apps (#256)
SumoLogic/sumologic-solution-templates: azure-collection-terraform/README.md

#### Provider deletion safety

This module exposes a boolean variable named `prevent_deletion_if_contains_resources` which controls the azurerm provider feature `resource_group.prevent_deletion_if_contains_resources`.

- Default: `true` (safe for production) — the provider will prevent deletion of a Resource Group if it still contains nested resources.
- For automated integration tests you may set this to `false` so the tests can clean up Resource Groups that still contain nested Azure resources (test-only usage).

Example (override via CLI):

```
terraform apply -var 'prevent_deletion_if_contains_resources=false'
```

Or in a tfvars file:

```hcl
prevent_deletion_if_contains_resources = false
```

#### Important Notes

**Activity Logs - Subscription-Level Warning:**

Activity logs operate at the **Azure subscription level**, not at the resource level. This has important implications:

- **Scope**. Activity logs capture ALL operations across the ENTIRE subscription (all resource groups, all resources).
- **Single Configuration**. Azure only allows **ONE** diagnostic setting for activity logs per subscription.

> - **Shared Resource**. If multiple Terraform configurations or users enable activity logs on the same subscription, **only the last one applied will be active**.

Reviewer: ```suggestion
- **Shared Resource**: If multiple Terraform configurations or users enable activity logs on the same subscription, **only the last one applied will be active**.
```

Reviewer: Can we change for Scope, Single Configurations and Deletion Impact as well replace "." with ":"

**Alternative Approach:**
If multiple teams need activity logs, consider:
- Creating a dedicated Terraform configuration for subscription-level resources (activity logs, policies, etc.)
- Using `terraform import` to manage existing activity log settings
- Documenting ownership and avoiding duplicate configurations

## How to Run This Project

### Step 1: Clone and Navigate

```bash
cd azure-collection-terraform
```

### Step 2: Create Configuration File

Copy the `terraform.tfvars.example` file to `terraform.tfvars` and configure the variables as described below:

```bash
cp terraform.tfvars.example terraform.tfvars
```

Edit `terraform.tfvars` and configure variables according to these groups:

#### 🔴 **Required - Must Configure** (Set these values for your environment)

```hcl
# ============================================================================
# Azure Authentication - Get these from Azure Portal
# ============================================================================
azure_subscription_id = "your-azure-subscription-id"  # Your Azure subscription ID
azure_client_id       = "your-azure-client-id"        # Service principal client ID

> azure_client_secret   = "your-azure-client-secret"    # Service principal secret

Reviewer: How can these be found by the user, the subscription ID and Tenant ID are easy to find but client ID and client secret we need to create service principal in that the client secret is created. Should we mention that or a link of how to do so?

In the above section in Config variables we only have explained what it is and not how can we create it

The `TARGET_RESOURCE_TYPES` variable supports multiple formats:

```bash
# JSON Array Format (Recommended)
TARGET_RESOURCE_TYPES=["Microsoft.KeyVault/vaults","Microsoft.Storage/storageAccounts"]

# Comma-separated Format
TARGET_RESOURCE_TYPES=Microsoft.KeyVault/vaults,Microsoft.Storage/storageAccounts

# Single Resource Type
TARGET_RESOURCE_TYPES=["Microsoft.KeyVault/vaults"]
```

#### Test Control Flags
```bash
RUN_INTEGRATION_TESTS=true    # Enable integration tests
CLEANUP_RESOURCES=true        # Auto-cleanup test resources
```

### Test Features

- **✅ Unified Resource Type Parsing**: Tests validate flexible parsing of `TARGET_RESOURCE_TYPES`
- **✅ Diagnostic Setting Naming**: Tests ensure Azure naming compliance with SumoLogic conventions
- **✅ Backward Compatibility**: Tests verify helper methods maintain compatibility with legacy variable names
- **✅ Environment Validation**: Tests validate all required environment variables and configurations
- **✅ Azure Resource Discovery**: Tests validate dynamic discovery of Azure resources
- **✅ SumoLogic Integration**: Tests validate proper SumoLogic collector and source configuration

## Troubleshooting

> ### Azure Event Hub Regional Support

Reviewer: I think mentioning the number of eventhubs instances different tiers i.e basic, standard etc can handle should be added in this section. The user may face errors if using standard and more than 10 instances are being created and will visit this section to troubleshoot.

> The table below shows Event Hub namespace availability across Azure regions by SKU tier. This module automatically handles region-specific SKU limitations.

Reviewer: ```suggestion
The tables below show Event Hub namespace availability across Azure regions by SKU tier. This module automatically handles region-specific SKU limitations.
```

| **Authentication failures with Azure provider** | 1. Ensure you've run `az login` and are authenticated<br>2. Verify your Azure credentials have sufficient permissions (Contributor + Monitoring Contributor roles)<br>3. Check subscription ID matches your target subscription<br>4. Run `az account show` to verify current context |
| **Variables validation errors** | 1. Check `eventhub_namespace_name` is 6-50 characters, starts with letter, globally unique<br>2. Verify `throughput_units` is one of: `1`, `2`, `4`, `8`, `16`<br>3. Ensure `location` matches a valid Azure region name<br>4. Validate `resource_group_name` doesn't contain special characters or spaces |

### Debugging

Enable detailed logs:
```bash
# Terraform debugging
export TF_LOG=DEBUG
terraform apply

# Azure CLI debugging
az account show --debug
```

Check EventHub metrics in Azure Portal:
- Monitor → Metrics → Select EventHub namespace
- View "Incoming Messages" and "Outgoing Messages"

## Cleanup

To destroy all resources:

```bash
terraform destroy
```

> **⚠️ Warning**: This will:

Reviewer: Can we move This will to next line ?? Same line has two colons :

# Azure Collection Terraform Module

Terraform module for automated log and metrics collection from Azure resources to Sumo Logic using EventHubs.

## Overview

This module creates a complete data pipeline to collect logs and metrics from Azure resources and send them to Sumo Logic for monitoring and analysis. It automatically:

- Discovers Azure resources based on tags
- Creates EventHub infrastructure per location
- Configures diagnostic settings for log collection

> - Sets up Sumo Logic collecto| <a name="output_sumo_metrics_sources"></a> [sumo\_metrics\_sources](#output\_sumo\_metrics\_sources) | A map of Sumo Logic metrics source IDs by namespace. |

Reviewer 2: ```suggestion
- Sets up Sumo Logic collector| <a name="output_sumo_metrics_sources"></a> [sumo\_metrics\_sources](#output\_sumo\_metrics\_sources) | A map of Sumo Logic metrics source IDs by namespace. |
```

> - Sets up Sumo Logic collecto| <a name="output_sumo_metrics_sources"></a> [sumo\_metrics\_sources](#output\_sumo\_metrics\_sources) | A map of Sumo Logic metrics source IDs by namespace. |

Reviewer 2: Can we check on the hyperlink used in this point. Also does it mean metric namespace

Author: link removed for now

> ## Troubleshooting

Reviewer 2: I think we should move troubleshooting and cleanup at the end of the doc

Author: let's not add any links for now

**Issue**: `EventHub throughput exceeded`
- **Solution**: Increase `eventhub_namespace_capacity` or upgrade to Premium SKU for higher throughput.

**Issue**: `Sumo Logic sources show "Not Receiving Data"`
- **Solution**: 
  1. Verify diagnostic settings are active in Azure Portal
  2. Check EventHub is receiving messages
  3. Confirm authorization rules have Listen permission
  4. Verify Sumo Logic connection string is correct

### Debugging

Enable detailed logs:
```bash
# Terraform debugging
export TF_LOG=DEBUG
terraform apply

# Azure CLI debugging
az account show --debug
```

Check EventHub metrics in Azure Portal:
- Monitor → Metrics → Select EventHub namespace
- View "Incoming Messages" and "Outgoing Messages"

## Cleanup

To destroy all resources:

```bash
terraform destroy
```

**Warning**: This will:
- Delete all EventHub namespaces and hubs
- Remove diagnostic settings from Azure resources
- Delete Sumo Logic collector and all sources
- Remove installed Sumo Logic apps

Data in Sumo Logic will be retained based on your retention policy.

> ## Architecture

Reviewer 2: Architecture could be moved to some internal doc ? I do not see value in giving this level of detail to user.

Author: let's keep as is for now

```
## Architecture

> ```

Reviewer 2: Can we add a line saying that this is what the data flow looks like

| [sumologic_azure_event_hub_log_source.sumo_azure_event_hub_log_source](https://registry.terraform.io/providers/SumoLogic/sumologic/latest/docs/resources/azure_event_hub_log_source) | resource |
| [sumologic_azure_metrics_source.terraform_azure_metrics_source](https://registry.terraform.io/providers/SumoLogic/sumologic/latest/docs/resources/azure_metrics_source) | resource |
| [sumologic_collector.sumo_collector](https://registry.terraform.io/providers/SumoLogic/sumologic/latest/docs/resources/collector) | resource |
| [azurerm_client_config.current](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/data-sources/client_config) | data source |
| [azurerm_monitor_diagnostic_categories.all_categories](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/data-sources/monitor_diagnostic_categories) | data source |
| [azurerm_resources.all_target_resources](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/data-sources/resources) | data source |

## Configuration

### Prerequisites

Before running this module, ensure you have:

1. **Azure**:
   - Active Azure subscription

>    - Service Principal with permissions:

Reviewer 2: Can we give a link here to let the user know how to create a service principal ? Or give some details of the steps invovled?

     - `Reader` on target resources
     - `Contributor` on resource group for EventHub creation
     - `Monitoring Contributor` for diagnostic settings

>    - Azure CLI configured or service principal credentials

Reviewer 2: Adding minimum roles required here will be helpful for the user

Author: --role "Contributor" is sufficient

2. **Sumo Logic**:
   - Active Sumo Logic account
   - Access ID and Access Key with permissions to create collectors and sources
   - Admin role recommended for app installation

3. **Terraform**:
   - Terraform >= 1.5.7 installed
   - Azure provider >= 4.19.0
   - Sumo Logic provider >= 3.0.2

### Key Variables

> #### Required Variables

Reviewer 2: Can we list in a tabular form, giving details of each input required. Similar to this : https://github.com/SumoLogic/sumologic-solution-templates/tree/master/aws-observability-terraform#inputs

```hcl
# Azure Authentication
azure_subscription_id = "your-subscription-id"
azure_client_id       = "your-client-id"
azure_client_secret   = "your-client-secret"
azure_tenant_id       = "your-tenant-id"

# Sumo Logic Authentication
sumologic_access_id      = "your-access-id"
sumologic_access_key     = "your-access-key"
sumologic_environment    = "us2"  # or your deployment region

# Resource Configuration
resource_group_name        = "rg-sumologic-eventhub"
location                   = "eastus"
collector_name             = "Azure Collection"
eventhub_namespace         = "sumo-logs"
```

#### Target Resources Configuration

Define which Azure resources to collect from using `target_resource_types`:

```hcl
target_resource_types = [
  {

>     log_namespace    = "Microsoft.Sql/servers/databases"

Reviewer 2: Do we need to take this from user ? Or should we maintain it ?

Author: we will mention all the namespaces in terraform.tfvars.example,  but user can update this going foreword

    metric_namespace = "Microsoft.Sql/servers/databases"
  },
  {
    log_namespace    = "Microsoft.Storage/storageAccounts"
    metric_namespace = "Microsoft.Storage/storageAccounts"
  }
]
```

**Fields:**
- `resource_type`: Azure resource type to discover
- `log_namespace`: Namespace for log collection (creates EventHub log source)
- `metric_namespace`: Namespace for metrics collection (creates metrics source)

**Nested Resources**: For parent-child resources (e.g., Storage sub-services), use `nested_namespace_configs`:

```hcl
nested_namespace_configs = {
  "Microsoft.Storage/storageAccounts" = [
    "Microsoft.Storage/storageAccounts/blobServices",
    "Microsoft.Storage/storageAccounts/fileServices"
  ]
}
```

#### App Installation

Install Sumo Logic apps for pre-built dashboards:

```hcl
installation_apps_list = [

>   {

Reviewer 2: can we make this generic, instead of hardcoded values of uuid name version etc. ? Since we just need to explain the structure of an object in the array

Author: we will add all apps with current versions in terraform.tfvars.example, will can update this going foreword

    uuid    = "0f2af8dd-447f-460f-95f7-3c7898a1eb25"
    name    = "Azure SQL"
    version = "1.0.0"
  },
  {
    uuid    = "c039e808-b5f4-4179-aba9-876c7eb01f94"
    name    = "Azure Storage"
    version = "1.0.0"
  }
]
```

#### Optional: Activity Logs

Enable subscription-level activity log collection:

```hcl
enable_activity_logs = true
```

## How to Run This Project

### Step 1: Clone and Navigate

```bash
cd azure-collection-terraform
```

### Step 2: Create Configuration File

> Create `terraform.tfvars` with your configuration:

Reviewer 2: ```suggestion
Create `terraform.tfvars` with your configuration. Here is a sample configuration:
```

  }
]

# Apps to Install
installation_apps_list = [
  {
    uuid    = "0f2af8dd-447f-460f-95f7-3c7898a1eb25"
    name    = "Azure SQL"
    version = "1.0.0"
  }
]

# Activity Logs
enable_activity_logs = true
```

### Step 3: Initialize Terraform

```bash
terraform init
```

This downloads required providers (Azure, Sumo Logic, Random, Time).

### Step 4: Review Plan

```bash
terraform plan
```

Review the resources to be created:
- Resource group for EventHub infrastructure
- EventHub namespaces (one per region)
- EventHubs (one per resource type and location)
- Diagnostic settings for each target resource
- Sumo Logic collector, sources, and apps

### Step 5: Deploy

```bash
terraform apply
```

Type `yes` when prompted. Deployment typically takes 5-10 minutes.

### Step 6: Verify

1. **Azure Portal**:
   - Check resource group contains EventHub namespaces
   - Verify diagnostic settings on target resources
   - Confirm EventHubs are receiving data

2. **Sumo Logic**:
   - Navigate to Manage Data → Collection
   - Verify collector is online
   - Check sources show "Receiving Data"
   - View installed apps in App Catalog

### Step 7: Access Dashboards

> Navigate to Sumo Logic dashboards for installed apps to view logs and metrics.

Reviewer 2: Can we be more specific - as to where specifically the content will be installed. Also we can mention about monitor creation using monitor template as part of installed apps.

## Outputs

After deployment, Terraform outputs:

| Output | Description |
|--------|-------------|
| `collector_id` | Sumo Logic collector ID |
| `eventhub_namespace_ids` | Map of EventHub namespace IDs by location |
| `eventhub_ids` | List of created EventHub IDs |
| `log_source_ids` | List of Sumo Logic log source IDs |
| `metrics_source_ids` | List of Sumo Logic metrics source IDs |
| `resource_group_id` | Azure resource group ID |
| `diagnostic_setting_ids` | Map of diagnostic setting IDs by resource |

> ## Testing

Reviewer 2: I think we should have a seperate INTERNAL_README.md which contains section like architecture and testing. Since those are developer specific info. This readme should address all the pointers of the user of this solution.

Author: for now kept single README, we will create separate one going foreword

Comprehensive test suite available in [`test/`](test/) directory. See [`test/README.md`](test/README.md) for:
- 20 validation tests covering Azure and Sumo Logic resources
- Integration test with 4-phase validation (infrastructure → configuration → data flow → cleanup)
- Test fixtures for various Azure resource types

**Quick Test:**
```bash
cd test
go test -v -run TestAzureCollectionValidation -timeout 30m
```

> ## Variable Reference (Auto-Generated)

Reviewer 2: Have we listed input and output twice - manually and auto-generated ? If there is not specific info in the manually generated one, we can keep the auto-generated once , since they are better structured.

| <a name="output_eventhub_namespaces_ids"></a> [eventhub\_namespaces\_ids](#output\_eventhub\_namespaces\_ids) | A map of Event Hub namespace IDs by location. |
| <a name="output_resource_group_name"></a> [resource\_group\_name](#output\_resource\_group\_name) | The name of the main resource group. |
| <a name="output_sumo_activity_log_source_id"></a> [sumo\_activity\_log\_source\_id](#output\_sumo\_activity\_log\_source\_id) | The ID of the Sumo Logic activity log source. |
| <a name="output_sumo_collection_policy_keys"></a> [sumo\_collection\_policy\_keys](#output\_sumo\_collection\_policy\_keys) | A map of shared access policy primary keys by location. |
| <a name="output_sumo_collector_id"></a> [sumo\_collector\_id](#output\_sumo\_collector\_id) | The ID of the Sumo Logic Hosted Collector. |
| <a name="output_sumo_eventhub_log_sources"></a> [sumo\_eventhub\_log\_sources](#output\_sumo\_eventhub\_log\_sources) | A map of Sumo Logic Event Hub log source IDs by resource type and location. |
| <a name="output_sumo_metrics_sources"></a> [sumo\_metrics\_sources](#output\_sumo\_metrics\_sources) | A map of Sumo Logic metrics source IDs by namespace. |

> ## Testing

Reviewer 2: Why is testing section mentioned twice ?

# Azure Collection Terraform Module

> Terraform module for automated log and metrics collection from Azure resources to Sumo Logic using EventHubs.

Reviewer 2: ```suggestion
Terraform module for automated log and metrics collection from Azure resources to Sumo Logic.
```

## Overview

This module creates a complete data pipeline to collect logs and metrics from Azure resources and send them to Sumo Logic for monitoring and analysis. It automatically:

- Discovers Azure resources based on tags
- Creates EventHub infrastructure per location
- Configures diagnostic settings for log collection

> - Sets up Sumo Logic collector

Reviewer 2: ```suggestion
- Sets up Sumo Logic sources and collectors
```

## Architecture

The following diagram illustrates the data flow architecture:

```

> Azure Resources → Diagnostic Settings → EventHubs → Sumo Logic Sources → Sumo Logic Apps

Reviewer 2: This only explain the flow for logs. We need to mention about the metrics collection as well. We can do that diving this into sections - logs and metrics

- **Configuration**: Specified via `installation_apps_list` variable
- **Examples**: Azure SQL, Azure Storage, Azure Kubernetes Service apps
- **Requirement**: Each app requires specific log/metric sources to be configured

### Data Sources

#### 1. **azurerm_client_config.current**
Retrieves current Azure authentication context.
- **Usage**: Gets subscription ID and tenant ID for resource tagging
- **Read-Only**: Does not create resources

#### 2. **azurerm_resources.all_target_resources**
Queries Azure for resources matching specified tags.
- **Purpose**: Discovers target resources for log collection
- **Filter**: Uses `resource_group_name_filter` and `resource_tags` variables
- **Output**: List of resource IDs, types, and locations
- **Dynamic**: Results determine EventHub and diagnostic setting creation

#### 3. **azurerm_monitor_diagnostic_categories.all_categories**
Retrieves available diagnostic categories for each discovered resource.
- **Purpose**: Determines what logs and metrics each resource can export
- **Dynamic**: Queries per discovered resource
- **Usage**: Ensures diagnostic settings enable all available log categories

> ## Requirements

Reviewer 2: Why is requirement specified twice ? The content under this heading do not make sense

> ```

Reviewer 2: This needs to be removed its corrupting the format

2. **EventHub Infrastructure**: Creates namespaces and hubs per location
3. **Diagnostic Settings**: Configures log streaming from resources to EventHubs
4. **Sumo Logic Integration**: Sets up collector, log/metric sources, and installs apps

## 📁 Configuration Files

- `terraform.tfvars.example` - Template with all variables and detailed documentation
- `terraform.tfvars` - Your actual configuration (customize and keep secure)
- `test/` - Comprehensive test suite with 69 test scenarios

## Requirements

| Name | Version |
|------|---------|
| <a name="requirement_terraform"></a> [terraform](#requirement\_terraform) | >= 1.5.7 |
| <a name="requirement_azurerm"></a> [azurerm](#requirement\_azurerm) | >= 4.19.0 |
| <a name="requirement_random"></a> [random](#requirement\_random) | >= 3.1.0 |
| <a name="requirement_sumologic"></a> [sumologic](#requirement\_sumologic) | >= 3.0.2 |
| <a name="requirement_time"></a> [time](#requirement\_time) | >= 0.7.1 |

## Providers

| Name | Version |
|------|---------|
| <a name="provider_azurerm"></a> [azurerm](#provider\_azurerm) | 4.43.0 |
| <a name="provider_sumologic"></a> [sumologic](#provider\_sumologic) | 3.1.4 |

## Modules

No modules.

> ## Resources

Reviewer 2: Why are we listing the resources twice. I feel the previous resource section seems to be too verbose at least from user prespective. We should just keep this one - tabular format - easy to read.

  --output table
```

#### 📋 **Authentication**

This module uses **Azure CLI Authentication**:
- Setup: Run `az login` before executing Terraform
- Terraform will use your Azure CLI session for all operations
- Best for: Local development and manual deployments

### Configuration Variables

The following table describes all available configuration variables. For a complete example, see `terraform.tfvars.example`.

| Name | Description | Type | Default | Required |
|------|-------------|------|---------|:--------:|
| **Azure Authentication** |||||
| `azure_subscription_id` | Azure Subscription ID where resources will be created. If not provided, uses current Azure CLI context. | `string` | `null` | Yes |
| `azure_client_id` | Service Principal Application (client) ID. If not provided, uses Azure CLI context. | `string` | `null` | Yes |
| `azure_client_secret` | Service Principal client secret value. If not provided, uses Azure CLI context. | `string` (sensitive) | `null` | Yes |
| `azure_tenant_id` | Azure Active Directory Tenant ID. If not provided, uses current Azure CLI context. | `string` | `null` | Yes |
| **Azure Infrastructure** |||||

> | `resource_group_name` | Name for the new resource group where all Event Hub namespaces will be created. These Event Hubs are used to configure resource logs and activity logs collection sources in Sumo Logic. | `string` | - | **Yes** |

Reviewer 2: Is there no default value for this ?

> | `eventhub_namespace_name` | Name for the Event Hub namespace (must be globally unique across Azure, 6-50 characters, starts with letter). | `string` | `"SUMO-267667-Hub"` | Yes |

Reviewer 2: Why is the default value SUMO-267667. Can we keep it to something meaningful
